LEGO1/tgl/d3drm/texture.cpp

```diff
@@ -0,0 +1,202 @@
+#include "impl.h"
+
+using namespace TglImpl;
+
+DECOMP_SIZE_ASSERT(TglD3DRMIMAGE, 0x40);
+
+inline TglD3DRMIMAGE* TextureGetImage(IDirect3DRMTexture* p_texture)
+{
+	return reinterpret_cast<TglD3DRMIMAGE*>(p_texture->GetAppData());
+}
+
+// Forward declare to satisfy order check
+void TextureDestroyCallback(IDirect3DRMObject* pObject, void* pArg);
+
+// OFFSET: LEGO1 0x100a12a0
+Result TextureImpl::SetImage(IDirect3DRMTexture* p_self, TglD3DRMIMAGE* p_image)
+{
+	unsigned long appData;
+	Result result;
+
+	appData = reinterpret_cast<unsigned long>(p_image);
+
+	// This is here because in the original code they asserted
+	// on the return value being NULL.
+	TextureGetImage(p_self);
+
+	result = ResultVal(p_self->SetAppData(appData));
+	if (Succeeded(result) && p_image) {
+		result = ResultVal(p_self->AddDestroyCallback(TextureDestroyCallback, NULL));
+		if (!Succeeded(result)) {
+			p_self->SetAppData(0);
+		}
+	}
+	return result;
+}
+
+// OFFSET: LEGO1 0x100a1300
+void TextureDestroyCallback(IDirect3DRMObject* pObject, void* pArg)
+{
+	TglD3DRMIMAGE* pImage = reinterpret_cast<TglD3DRMIMAGE*>(pObject->GetAppData());
+	delete pImage;
+	pObject->SetAppData(0);
+}
+
+// OFFSET: LEGO1 0x100a1330
+TglD3DRMIMAGE::TglD3DRMIMAGE(
+	int p_width,
+	int p_height,
+	int p_depth,
+	void* p_buffer,
+	int p_useBuffer,
+	int p_paletteSize,
+	PaletteEntry* p_palette
+)
+{
+	m_image.aspectx = 1;
+	m_image.aspecty = 1;
+	m_image.width = 0;
+	m_image.height = 0;
+	m_image.depth = 0;
+	m_image.rgb = 0;
+	m_image.bytes_per_line = 0;
+	m_image.buffer1 = NULL;
+	m_image.buffer2 = NULL;
+	m_image.red_mask = 0xFF;
+	m_image.green_mask = 0xFF;
+	m_image.blue_mask = 0xFF;
+	m_image.alpha_mask = 0xFF;
+	m_image.palette_size = 0;
+	m_image.palette = NULL;
+	m_texelsAllocatedByClient = 0;
+	if (p_buffer != NULL) {
+		CreateBuffer(p_width, p_height, p_depth, p_buffer, p_useBuffer);
+	}
+	if (p_palette != NULL) {
+		InitializePalette(p_paletteSize, p_palette);
+	}
+}
+
+// OFFSET: LEGO1 0x100a13b0
+void TglD3DRMIMAGE::Destroy()
+{
+	if (m_texelsAllocatedByClient == 0) {
+		free(m_image.buffer1);
```

Comment:
This should be `delete` (below too)

Reply:
This definitely can't be a new/delete allocation because there's no null check on the frees in the destructor.

Reply:
I might be missing something here, but if I replace both `free`'s in this function with `delete`, I'm getting the exact the same match/disassembly.

Could you elaborate where you see the difference exactly that ascertains that it must be `free`?

Reply:
No, you're right, I was just blundering. I got myself thinking that the ifs were the C++ `delete` adding an implicit test, but the `free` also has an embedded null check, so there's no reason for C++ to be doing that.